pyro/optim/multi.py

```diff
@@ -0,0 +1,114 @@
+from __future__ import absolute_import, division, print_function
+
+import torch
+
+from pyro.ops.newton import newton_step_2d
+from pyro.optim.optim import PyroOptim
+
+
+class MultiOptimizer(object):
+    """
+    Base class of optimizers that make use of higher-order derivatives.
+
+    Higher-order optimizers generally use :func:`torch.autograd.grad` rather
+    than :meth:`torch.Tensor.backward`, and therefore require a different
+    interface from usual Pyro and PyTorch optimizers. In this interface,
+    the :meth:`step` method inputs a ``loss`` tensor to be differentiated,
+    and backpropagation is triggered one or more times inside the optimizer.
+
+    Derived classes must implement :meth:`step` to compute derivatives and
+    update parameters in-place.
+
+    Example::
+
+        tr = poutine.trace(model).get_trace(*args, **kwargs)
+        loss = -tr.log_prob()
+        params = {name: pyro.param(name).unconstrained()
+                  for name in pyro.get_param_store().get_all_param_names()}
+        optim.step(loss, params)
+    """
+    def step(self, loss, params):
+        """
+        Performs an optimization step on parameters given a differentiable
+        ``loss`` tensor.
+
+        :param torch.Tensor loss: A differentiable tensor to be minimized.
+            Some optimizers require this to be differentiable multiple
+            times.
+        :param dict params: A dictionary mapping param name to unconstrained
+            value as stored in the param store.
+        """
+        raise NotImplementedError
+
+
+class PyroMultiOptimizer(MultiOptimizer):
+    """
+    Facade to wrap :class:`~pyro.optim.optim.PyroOptim` objects
+    in a :class:`MultiOptimizer` interface.
+    """
+    def __init__(self, optim):
+        self.optim = optim
+
+    def step(self, loss, params):
+        names = []
+        values = []
+        for name, value in params.items():
+            names.append(name)
+            values.append(value)
+        grads = torch.autograd.grad(loss, values, create_graph=True)
+        for x, g in zip(values, grads):
+            x.grad = g
```

Comment:
I think that the usual behavior of a Pytorch `optimizer` accumulates gradients, not set `.grad` like this. How about use [torch.autograd.backward](https://pytorch.org/docs/master/autograd.html#torch.autograd.backward) here instead? Something like this:
```
grads = torch.autograd.grad(loss, values, create_graph=True)
torch.autograd.backward(values, grads)
```

Reply:
I think we cannot use `.backward()` because backward operates on the entire compute graph reachable (upstream from) the `loss` tensor, whereas we want to restrict to computing gradients only wrt the given variables. This is important for interleaving multiple optimizers via `MixedMultiOptimizer`. Does that seem correct to you?

Reply:
I think that `torch.autograd.backward(values, grads)` just operates on the param values, not the entire graph. According to Adam in Pytorch's slack autograd channel (you can search about `torch.autograd.backward` there), this method is slightly faster.

Reply:
`torch.autograd.backward(values, grads)` "Computes the sum of gradients of given tensors w.r.t. graph leaves.", whereas we want to compute gradients only w.r.t. a subset of the leaves. Thus `backward` is not compatible with `MixedMultiOptimizer`. Adam has commented that `grad` is negligibly faster than `backward`; it performs the same tensor ops and has to walk the autograd graph twice, however this walking is cheap.

Reply:
@fritzo I think that in this case, these `values` are leaves, so this method just accumulates the `grads` into their `.grad` attributes (no real `backward` is processed). However, I agree that if your purpose is to not accumulate (`x.grad = g` instead of `x.grad = x.grad + g`), then we don't need to use that hacking statement.

Reply:
I believe the `values` are not leaves but rather roots. By default `values = [loss]` and `grads = [torch.tensor(1.)]`

Reply:
I mean `values` and `grads` are `values` and `grads` in this statement `grads = torch.autograd.grad(loss, values, create_graph=True)` and these `values` are params. The code I have mentioned is
```
grads = torch.autograd.grad(loss, values, create_graph=True)
torch.autograd.backward(values, grads)  # not torch.autograd.backward([loss], [torch.tensor(1.)])
```
Sorry if I still miss something.